### ISSUE TITLE
More complex signatures appear to cause a drop out on Blazer Server

### ISSUE THREAD

**Issue by michaelfw2020** [Mobsites/Blazor.SignaturePad#2]
Hi Allan,

Loving the component, one thing we have noticed in testing, is when you use a simple signature everything appears great. When you have a slightly more complex signature, we get a disconnect from server end error.

This occurs on:-

dataURL = await signaturePad.ToDataURL(SignaturePad.SupportedImageTypes.jpg);

Tried Catching the error to no avail, in browser console we see:-

blazor.server.js:1 [2020-04-22T12:59:34.806Z] Error: Connection disconnected with error 'Error: Server returned an error on close: Connection closed with an error.'.

Thanks
Mike

**Comment by allan-mobley-jr** [comment]
Okay, thanks!

Currently preparing an official release out of preview. So this is perfect timing. I will look into this and see what I can come up with.

**Comment by michaelfw2020** [comment]
Thanks Allan, One thing I have just done is to monitor the string length returned up until it starts to fail, it gets up to about about 32,500 ish, then anything above this seems to fail. Thought it may help. Thanks. Mike

**Comment by allan-mobley-jr** [comment]
I was wondering if it were a memory issue. This may be surpassing a limit imposed by Blazor server-side via circuits.

**Comment by michaelfw2020** [comment]
No you mention that. I was reading up on it and it does suggest a message size limit of 32k on the circuits.

**Comment by allan-mobley-jr** [comment]
Can you post a screenshot of the signature on the pad when this occurs? Or does the app crash making it impossible to capture the screen?

**Comment by michaelfw2020** [comment]
Ok up until this point:-

![image](https://user-images.githubusercontent.com/64143194/79992752-d2417280-84ab-11ea-9b14-e08e0f27126f.png)

**Comment by michaelfw2020** [comment]
Then fails here:

![image](https://user-images.githubusercontent.com/64143194/79992838-ee451400-84ab-11ea-8376-2bfadec043e3.png)

**Comment by allan-mobley-jr** [comment]
Thanks!

**Comment by allan-mobley-jr** [comment]
Mike,

Looking into this, it appears that there needs to be a special case for handling the return data from the toDataURL() method when Blazor Server is being used.

Perhaps, a separate method or a parameter flag to inform C# and javascript that the data needs to be sent in chunks less than the limit.

At any rate, the component or page that consumes Blazor Signature Pad would have to loop the separate method until all the chunks are fetched.

Not sure of the exact implementation yet, but this is what I am thinking.

Any thoughts?

Thanks.

**Comment by michaelfw2020** [comment]
Hi Allan,

Yes I agree. Did some further reading this afternoon and found this page from MS.

[https://docs.microsoft.com/en-us/aspnet/core/blazor/advanced-scenarios?view=aspnetcore-3.1](url)

Appears to do as you suggest, breaks the return data down into segments.

Thanks
Mike

**Comment by michaelfw2020** [comment]
Hi Allan,

Also, it appears the limitation exists between JS and Blazor C#. So I am thinking maybe it could be modified in SigaturePad.razor.cs. On this section:-

/// <summary>
        /// Get signature as data url according to the supported type.
        /// </summary>
        public async ValueTask<string> ToDataURL(SupportedImageTypes type) => 
            await jsRuntime.InvokeAsync<string>(
                "Blazor.SignaturePad.toDataURL",
                type.ToString());

        /// <summary>
        /// Invoked from a javascript callback event when a signature stroke finishes or is removed, or when the signature is cleared.
        /// </summary>

So the JS is updated in line with MS suggestion with a GetFile size and GetSegements function, and this function above is also adjusted to handle the looping to read the GetFileSize and GetSegments from the JS.

That way then, end users consuming your component can still continue to use as current and not have to worry about the looping and Segments.

But that is based on an assumption at this point.

Thanks
Mike

**Comment by allan-mobley-jr** [comment]
Yes, that sounds like a good approach.

Still, I'm not sure this needs to be enforced for Blazor wasm. It seems to handle everything I have thrown at it within the realm of possibility.

Perhaps, an attribute flag from the user that the Signature Pad is being used in Blazor Server?

<SignaturePad IsBlazorServer></SignaturePad>

But I see your point about fixing it this way no matter whether server or wasm, as then the user would not have to do anything.

**Comment by michaelfw2020** [comment]
Hi Allan,

Still quite new to this Blazor side of things. But am loving it.

You’ve done a cracking job on this. First time I’ve installed one, followed the instructions and it worked out of the tin first time. 

Yes could add a flag or like you say, implement as standard so it’s agnostic to the consumer. Not really looked at wasm yet. Next on the list. 

If you are stacked with work at the moment, I can have a little play locally tomorrow and see where I get to with it?

Thanks
Mike

**Comment by allan-mobley-jr** [comment]
Sure. I would like that very much.

Check out the v.1.0.0 branch. I will push my latest commits here shortly. There is a lot of changes from the preview. It's incorporating a base component library that is being used in other components that I am working on as well.

Thanks Mike!

**Comment by allan-mobley-jr** [comment]
Mike,

Just pushed the latest for branch v.1.0.0. That's everything for the official release pending what we come up with on this issue and a few minor tweaks I might make with styles.

**Comment by allan-mobley-jr** [comment]
You should get an invite now. GitHub was just having major issues...at least for me.

**Comment by allan-mobley-jr** [comment]
Create a pull request

**Comment by michaelfw2020** [comment]
Hi Allan,

I pushed my changes and it got further. But got this:-

Remote: Resolving deltas: 100% (5/5), completed with 5 local objects.        
Remote: error: GH006: Protected branch update failed for refs/heads/v1.0.0.        
Remote: error: At least 1 approving review is required by reviewers with write access.        

Mean anything to you ?

Cheers
Mike

**Comment by allan-mobley-jr** [comment]
Bear with me. I am new to the whole collaborator-thing and open source.

Did you try creating a pull request?

**Comment by michaelfw2020** [comment]
Yes, came back with Repository up to date. No Changes.

**Comment by allan-mobley-jr** [comment]
Check this [article](https://codeburst.io/a-step-by-step-guide-to-making-your-first-github-contribution-5302260a2940) out.

I think you have to fork it first.

**Comment by michaelfw2020** [comment]
Hi Allan,

So I forked it and it appears to have pushed a new branch to you, I hope you can see that. Will that now let you check it and then Merge ?

Thanks
MIke

**Comment by allan-mobley-jr** [comment]
Yes. I will check out what you got. Go ahead and try the pull request to have it merged with v1.0.0.0.

Thanks Mike!

**Comment by michaelfw2020** [comment]
No problems. Let me know how you get on and if its ok.

**Comment by allan-mobley-jr** [comment]
Will do

**Comment by allan-mobley-jr** [comment]
Mike,

That works great!

Couple of things:
* Do we integrate this functionality into the original method and use this as the default handling for both wasm and server?
* If not, do we create an attribute flag for the user that they set on the Signature Pad tag and then automatically call the server version (making it an internal method), or do we leave it as is and have the user call this when using Blazor server and not the other.

**Comment by michaelfw2020** [comment]
HI Allan,

Personally I think it would be good to just replace the old with the new and keep one function, which should be transparent to the end user. I left in both so you can compare and check with the wasm, as I have never used this. 

Thanks
Mike

**Comment by allan-mobley-jr** [comment]
Okay, let me run this on wasm. 

If all is good, I'll have you incorporate this into the original.

Also, you left one js function outside of the Signature Pad object. Can you move it the object like the others? This is just in case someone who uses this library doesn't already have a method with the same name.

Thanks so much for your time and help. Excellent!

**Comment by michaelfw2020** [comment]
Hi Allan,

I made the change to the app.js. It made me branch again. So you will see another branch in there.

Thanks
MIke

**Comment by allan-mobley-jr** [comment]
Sorry about that. You should be able to update your original branch now. My pull request restriction was too broad.

**Comment by allan-mobley-jr** [comment]
Blazor wasm seems to work just fine with what you have.

Let's move this into the original method.

Make all your changes to your original branch.

Then create a pull request to have it merged into v1.0.0

And I will merge it.

Awesome job, Mike!

**Comment by michaelfw2020** [comment]
Hi Allan,

Have merged into original branch and pushed back up.

So you should have it now.

Thanks
Mike

**Comment by allan-mobley-jr** [comment]
Mike,

It appears that when I did a pull request to merge your fix for Blazor server it triggered all of my yml pipelines and created a nuget package for v1.0.0.

This was a surprise to me as I thought my pipeline build and release were tied only to the master branch. Yet when I merged your branch into branch v1.0.0, it nonetheless triggered the runs lol.

I just made some final adjustments for saving the signature in browser storage when keeping state is enabled. I moved your fix into the original method and was going to call it a day and push the official release only to find that it had already occurred.

So I will have to create another release, say v1.1.0 since nuget will not let me roll back a package.

But first, I need to find out why this happened. I don't know why a pull request would cause this to happen by default. Totally unexpected and quite frankly stupid.
